src/core/main/common/FreezeResolver.ts

```diff
@@ -0,0 +1,573 @@
+          // seconds of playback behind - we don't want to compare things
+          // that happended too long ago.
+          freezingPosition - segment.position < 3000
+        ) {
+          // We're still playing the last segment at that point, update it.
+          //
+          // (We may be playing, or be freezing, on the current segment for some
+          // time, this allows to consider a more precize timestamp at which we
+          // switched segments).
+          currentSegmentEntry = segment;
+        }
+      }
+
+      if (
+        previousRepresentationEntry === undefined ||
+        previousRepresentationEntry.segment === null
+      ) {
+        log.debug(
+          "FR: Freeze when beginning to play a content, try avoiding this quality",
+        );
+        toAvoid.push({
+          adaptation: currentSegment.infos.adaptation,
+          period: currentSegment.infos.period,
+          representation: currentSegment.infos.representation,
+        });
+      } else if (
+        currentSegment.infos.period.id !==
+        previousRepresentationEntry.segment.infos.period.id
+      ) {
+        log.debug("FR: Freeze when switching Period, reloading");
+        return { type: "reload", value: null };
+      } else if (
+        currentSegment.infos.representation.uniqueId !==
+        previousRepresentationEntry.segment.infos.representation.uniqueId
+      ) {
+        log.warn(
+          "FR: Freeze when switching Representation, avoiding",
+          currentSegment.infos.representation.bitrate,
+        );
+        toAvoid.push({
+          adaptation: currentSegment.infos.adaptation,
+          period: currentSegment.infos.period,
+          representation: currentSegment.infos.representation,
+        });
+      }
+    }
+
+    if (toAvoid.length > 0) {
+      return { type: "avoid-representations", value: toAvoid };
+    } else {
+      log.debug("FR: Reloading because flush doesn't work");
+      return { type: "reload", value: null };
+    }
+  }
+
+  /**
+   * Add entry to `this._lastSegmentInfo` for the position that is currently
+   * played according to the given `observation`.
+   *
+   * @param {Object} observation
+   * @param {number} currentTimestamp
+   */
+  private _addPositionToHistory(
+    observation: IFreezeResolverObservation,
+    currentTimestamp: number,
+  ): void {
+    const position = observation.position.getPolled();
+    for (const ttype of ["audio", "video"] as const) {
+      const status = this._segmentSinksStore.getStatus(ttype);
+      if (status.type === "initialized") {
+        for (const segment of status.value.getLastKnownInventory()) {
+          if (
+            (segment.bufferedStart ?? segment.start) <= position &&
+            (segment.bufferedEnd ?? segment.end) > position
+          ) {
+            this._lastSegmentInfo[ttype].push({
+              segment,
+              position,
+              timestamp: currentTimestamp,
+            });
```

Comment:
I think this would miss some segments if the Observation interval is greater than the segment size,  no?

Reply:
You're right, it would thus end up not pinpointing the exact time at which a quality transition has been done, and I guess the effect would be limited, but that part of the code does not know about that.

What do you propose we do?

Reduce the observation's interval? (Note: I would like to do that at some point because it's not the first time where I would prefer more frequent updates).

More complex checks not only based on the current position but the range of positions since the last tick?

Add comments to indicate that this history might miss small segments?